fetch/fetch.go

```diff
@@ -0,0 +1,526 @@
+	success          HandleDataCallback
+	hash             types.Hash32
+	priority         priority
+	validateResponse bool
+	hint             Hint
+	returnChan       chan HashDataPromiseResult
+}
+
+// requestMessage is the on the wire message that will be send to the peer for hash query
+type requestMessage struct {
+	Hint Hint
+	Hash types.Hash32
+}
+
+// responseMessage is the on the wire message that will be send to the this node as response,
+type responseMessage struct {
+	Hash types.Hash32
+	data []byte
+}
+
+// requestBatch is a batch of Requests and a hash of all Requests as ID
+type requestBatch struct {
+	ID       types.Hash32
+	Requests []requestMessage
+}
+
+// SetID calculates the hash of all Requests and sets it as this batches ID
+func (b *requestBatch) SetID() {
+	bts, err := types.InterfaceToBytes(b.Requests)
+	if err != nil {
+		return
+	}
+	b.ID = types.CalcHash32(bts)
+}
+
+//ToMap converts the array of Requests to map so it can be easily invalidated
+func (b requestBatch) ToMap() map[types.Hash32]requestMessage {
+	m := make(map[types.Hash32]requestMessage)
+	for _, r := range b.Requests {
+		m[r.Hash] = r
+	}
+	return m
+}
+
+// responseBatch is the response struct send for a requestBatch. the responseBatch ID must be the same
+// as stated in requestBatch even if not all data is present
+type responseBatch struct {
+	ID        types.Hash32
+	Responses []responseMessage
+}
+
+// Config is the configuration file of the Fetch component
+type Config struct {
+	BatchTimeout      int
+	MaxRetiresForPeer int
+	BatchSize         int
+	RequestTimeout    int
+}
+
+type peers interface {
+	GetPeers() []p2ppeers.Peer
+}
+
+// MessageNetwork is a network interface that allows fetch to communicate with other nodes with 'fetch servers'
+type MessageNetwork struct {
+	*server.MessageServer
+	peers
+	log.Log
+}
+
+// NewMessageNetwork creates a new instance of the fetch network server
+func NewMessageNetwork(requestTimeOut int, net service.Service, protocol string, log log.Log) *MessageNetwork {
+	return &MessageNetwork{
+		server.NewMsgServer(net.(server.Service), protocol, time.Duration(requestTimeOut)*time.Second, make(chan service.DirectMessage, p2pconf.Values.BufferSize), log),
+		p2ppeers.NewPeers(net, log.WithName("peers")),
+		log,
+	}
+}
+
+// GetRandomPeer returns a random peer from current peer list
+func (f MessageNetwork) GetRandomPeer() p2ppeers.Peer {
+	peers := f.peers.GetPeers()
+	if len(peers) == 0 {
+		f.Log.Panic("cannot send fetch - no peers found")
+	}
+	rand.Seed(time.Now().Unix()) // initialize global pseudo random generator
+	return peers[rand.Intn(len(peers))]
+}
+
+type network interface {
+	GetRandomPeer() p2ppeers.Peer
+	SendRequest(msgType server.MessageType, payload []byte, address p2pcrypto.PublicKey, resHandler func(msg []byte), failHandler func(err error)) error
+}
+
+// Fetch is the main struct that contains network peers and logic to batch and dispatch hash fetch Requests
+type Fetch struct {
```

Comment:
It will be replaced by syncer, currently it is not connected to actual sync so it's under a different name

Reply:
It shouldn't be called`Fetcher` just because the current `Syncer` isn't going to be removed right away, because `Fetcher` implies that it's just about requesting data, while `Syncer` does better job in implying that it is for listening/serving other nodes requests' as well. Also, this makes it unclear that this is ought to replace the current `Syncer`. 

So in the meanwhile consider calling it `Syncer2`, `TurboSyncer`, etc.

Reply:
It shouldn't be called`Fetcher` just because the current `Syncer` isn't going to be removed right away, because `Fetcher` implies that it's just about requesting data, while `Syncer` does better job in implying that it is for listening/serving other nodes requests' as well. Also, this makes it unclear that this is ought to replace the current `Syncer`. 

So in the meanwhile consider calling it `Syncer2`, `TurboSyncer`, etc.

Reply:
I think we'll stick to fetcher for now... since this component does not really perform sync logic - it is much simpler - more only server basic fetch